### PR TITLE
Remove scheduled routing

### PR DESCRIPTION
**Pull request by kimchy**
Today, we have scheduled reroute that kicks every 10 seconds and checks if a
reroute is needed. We use it when adding nodes, since we don't reroute right
away once its added, and give it a time window to add additional nodes.

We do have recover after nodes setting and such in order to wait for enough
nodes to be added, and also, it really depends at what part of the 10s window
you end up, sometimes, it might not be effective at all. In general, its historic
from the times before we had recover after nodes and such.

This change removes the 10s scheduling, simplifies RoutingService, and adds
explicit reroute when a node is added to the system. It also adds unit tests
to RoutingService.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/test/InternalTestCluster.java [hunk 1]
```diff
@@ -365,7 +365,6 @@ private static Settings getRandomNodeSettings(long seed) {
         Random random = new Random(seed);
         Builder builder = Settings.settingsBuilder()
                 // decrease the routing schedule so new nodes will be added quickly - some random value between 30 and 80 ms
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/test/InternalTestCluster.java]: comment can go away..

File: core/src/test/java/org/elasticsearch/cluster/routing/RoutingServiceTests.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        routingService = new TestRoutingService();
+    }
+
+    @After
+    public void shutdownRoutingService() throws Exception {
+        routingService.shutdown();
+    }
+
+    @Test
+    public void testReroute() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/RoutingServiceTests.java]: I'm confused by this test, what does it do?

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java [hunk 3]
```diff
@@ -90,52 +79,26 @@ protected void doStop() {
...
-        FutureUtils.cancel(scheduledRoutingTableFuture);
-        scheduledRoutingTableFuture = null;
         clusterService.remove(this);
     }
 
-    /** make sure that a reroute will be done by the next scheduled check */
-    public void scheduleReroute() {
-        routingTableDirty = true;
+    /**
+     * make sure that a reroute will be done by the next scheduled check
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: comment probably needs to be adapted - it's done immediately now,

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java [hunk 4]
```diff
@@ -90,52 +79,26 @@ protected void doStop() {
...
+     */
+    public void reroute(String reason) {
+        performReroute(reason);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: do we need the performReroute? can't we fold this here?

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java [hunk 5]
```diff
@@ -158,25 +120,26 @@ public void onFailure(Throwable t) {
...
         try {
-            if (!routingTableDirty) {
-                return;
-            }
             if (lifecycle.stopped()) {
                 return;
             }
             if (rerouting.compareAndSet(false, true) == false) {
-                logger.trace("already has pending reroute, ignoring");
+                logger.trace("already has pending reroute, ignoring {}", reason);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: Just an idea how about making rerouting an AtomicReference<String> where a non null value signals the currently next reroute task name? this way we can log here logger.trace("already has pending reroute due to [{}]. ignoring [{}]")/

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java [hunk 6]
```diff
@@ -932,7 +932,10 @@ public ClusterState execute(ClusterState currentState) {
                     if (modified) {
                         stateBuilder.nodes(nodesBuilder);
                     }
-                    return stateBuilder.build();
+                    currentState = stateBuilder.build();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: I wonder if we should shortcut if modified == false . I'm thinking about a join storm upon master failure which will lead to a series of "received a join request for an existing node " where we don't really need to do anything.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java [hunk 7]
```diff
@@ -90,52 +79,26 @@ protected void doStop() {
 
     @Override
     protected void doClose() {
-        FutureUtils.cancel(scheduledRoutingTableFuture);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: not related to this change, but I think it's good to do FutureUtils.cancel(registeredNextDelayFuture) here. Might save on some cryptic error messages during shutdown..

**Comment by bleskes** [comment]
I think it's awesome how easy this turned up to be. I looked at all the places we submit a cluster state update task and didn't see any place we miss a reroute (and would rely on the 10s) apart from the node join we already knew about. 

Made some small suggestions.

File: core/src/test/java/org/elasticsearch/test/InternalTestCluster.java (hunk 1 above)
Comment (kimchy) [inline core/src/test/java/org/elasticsearch/test/InternalTestCluster.java]: right, missed it..

File: core/src/test/java/org/elasticsearch/cluster/routing/RoutingServiceTests.java (hunk 2 above)
Comment (kimchy) [inline core/src/test/java/org/elasticsearch/cluster/routing/RoutingServiceTests.java]: check that reroute actually triggers the reroute flag...

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java (hunk 3 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: right, will amend.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java (hunk 4 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: I wanted to separate the method we override in our tests from the public method, will make this one final

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java (hunk 6 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: I was looking to reduce the amount of logical change here, but I can make this change...

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java (hunk 7 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: I am not sure its a real problem, cause on shutdown we cancel ongoing ones anyhow, but will add

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java [hunk 8]
```diff
@@ -76,8 +64,9 @@ public RoutingService(Settings settings, ThreadPool threadPool, ClusterService c
         this.threadPool = threadPool;
         this.clusterService = clusterService;
         this.allocationService = allocationService;
-        this.schedule = settings.getAsTime("cluster.routing.schedule", timeValueSeconds(10));
-        clusterService.addFirst(this);
+        if (clusterService != null) {
+            clusterService.addFirst(this);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java]: at some point we should really add these ourside of the classes itself IMO. but that is totally unrelated

File: core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 9]
```diff
@@ -87,9 +83,8 @@ public GatewayAllocator(Settings settings, TransportNodesListGatewayStartedShard
         logger.debug("using initial_shards [{}]", initialShards);
     }
 
-    public void setReallocation(final ClusterService clusterService, final AllocationService allocationService) {
-        this.clusterService = clusterService;
-        this.allocationService = allocationService;
+    public void setReallocation(final ClusterService clusterService, final RoutingService routingService) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: I wonder if there is still a circular dep. now that AllocatinService is not needed here?

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java (hunk 6 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: actually, I don't want to change this behavior to be honest, I am going to leave it as is. We might to actually do want to send another cluster state as a result

File: core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java (hunk 9 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: I also wonder if we can trash this method then altogether a make `GatewayAllocator implement ClusterStateListener` and add it to the `ClusterService` once constructed or maybe inside the ClusterService...

**Comment by kimchy** [comment]
@bleskes I pushed a first review round, I don't like the suggestion on the atomic reference one, I don't think it adds a lot of value, on the other hand, I have an idea on what would, but its a bigger change, will see if I can work on it and if it stays small

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: We do need to publish a cluster state to seal the mastership on the joining nodes. I will add a big fat comment about this here after this is committed (to not be in your way when you merge). What I meant is avoiding the reroute, which was accumulated before for all the joins. But ok to leave as is for now if you have concerns.

File: core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java (hunk 9 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: the cycling problem will still remain, do you mean fold RoutingService into ClusterService? I kinda like that ClusterService has no notion or knowledge of allocation / routing, and just deals with cluster state in a generic form

**Comment by bleskes** [comment]
LGTM. I'm +1 on 1.7 - it makes things easier to trace, reason about and debug.

File: core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java (hunk 9 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: oh man that class is a nighmare. really I just realized how fucked up this is grrr.

**Comment by s1monw** [comment]
I am ok with this too... this entire circular dep stuff is just really bad but lets move on
